File: pi_appx.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 31 12:22:02 2020

@author: Ahmad

# Pi approximation using Monte Carlo simulation
"""

import random
import math

def distance(point1_coords, point2_coords):
    '''  returns distance between two points:  distance = sqrt( (x2-x1) + (y2-y1) )  '''
    
    x1, y1 = point1_coords[0], point1_coords[1]
    x2, y2 = point2_coords[0], point2_coords[1]
    
    return ((x2-x1)**2 + (y2-y1)**2)**0.50


def point_in_circle(dist, radius):
    '''  returns True if point is inside the circle, False otherwise  '''

    inside_circle = False

    if dist<=radius:
        inside_circle = True
    
    return inside_circle


if __name__ == '__main__':
    origin = (0,0)
    circle_radius = 1
    square_area = 4

    random_sub_add = [-1, 1]
    
    #random_point = (round(random.choice(random_sub_add)*random.random(), 4), round(random.choice(random_sub_add)*random.random(), 4))
    #point_dist = round(distance(origin, random_point), 4)   
    #in_circle = point_in_circle(point_dist, circle_radius)
    
    # monte-carlo simulation
    simulations = 1001
    num_points_in_circle = 0

    for i in range(1,simulations):
        # obtain a random point within the square
        random_point = (round(random.choice(random_sub_add)*random.random(), 4), round(random.choice(random_sub_add)*random.random(), 4))
        
        # calculate its distance from the center
        point_dist = round(distance(origin, random_point), 4)
        
        # is the point inside or outside the circle
        in_circle = point_in_circle(point_dist, circle_radius)
        
        # if inside, increment the counter for points inside the circle
        if in_circle:
            num_points_in_circle += 1
        
        # approximate PI by calculating percentage of square area in which points fall inside the circle 
        pi_appx = num_points_in_circle/i * square_area

        print(f'Simulation {i}:  random point={random_point},   distance={point_dist},   points inside circle={num_points_in_circle},  PI appx={pi_appx}')
        
    
    print(f'\nFinal PI approximation = {pi_appx}')

File: pi_appx.cpp
/******************************************************************************
Created on Sat Mar 27 12:22:02 2020

@author: Ahmad

# Pi approximation using Monte Carlo simulation
*******************************************************************************/

#include <iostream>
#include<tuple>
#include <string> 
#include <cmath> 

using namespace std;

float gen_random_number(int min, int max) {
    float random_num;
    random_num = min + static_cast <float> (rand()) /( static_cast <float> (RAND_MAX/(max-min)));
    return(random_num);
}

float euclidean_distance(tuple <float, float> point1_coords, tuple <float, float> point2_coords) {
    float distance;
    float x1, x2, y1, y2;
    
    x1 = get<0>(point1_coords); y1 = get<1>(point1_coords);
    x2 = get<0>(point2_coords); y2 = get<1>(point2_coords);
    distance = pow(((x2-x1)*(x2-x1) + (y2-y1)*(y2-y1)), 0.50);
    
    return(distance);
}

bool point_in_circle(float dist, float radius) {
    bool inside_circle = false;
    
    if (dist<=radius)
        inside_circle = true;
        
    return(inside_circle);
}


int main()
{
	// defining origin and random point
    tuple <float, float> origin=make_tuple(0.0,0.0); 
    tuple <float, float> rand_point;
	
	// defining the circle of radius 1 unit, square with sides of 2 units
    float circle_radius=1, square_side=2, square_area, random_num, distance, pi_value;
	int simulations=10001, circle_counter=0, min=-1, max=1;
    bool in_circle;

    square_area = square_side * square_side;
    
    for (int i = 0; i < simulations; i++) {

		// obtain random (x,y) point within the square
        get<0>(rand_point) = gen_random_number(min, max);
        get<1>(rand_point) = gen_random_number(min, max);
		
		// get distance of the random point from the origin
        distance = euclidean_distance(origin, rand_point);
		
		// if the point is inside the circle, increment circle_counter
        in_circle = point_in_circle(distance, circle_radius);
        
        if (in_circle)
            circle_counter += 1;
            
		// approximate pi by multiplying ratio of points inside circle to the area of square
        pi_value = (circle_counter *1.0/ i) * square_area;
        cout<<pi_value<<endl;
    }

    return 0;
}